File: slp/backend-dotnet/Features/Quiz/QuizService.cs
using backend_dotnet.Features.Helpers;
using backend_dotnet.Features.Note;
using backend_dotnet.Features.Source;
using backend_dotnet.Features.Tag;
using backend_dotnet.Helpers;          
using System.Text.Json;

namespace backend_dotnet.Features.Quiz;

public class QuizService : IQuizService
{
    private readonly IQuizRepository _quizRepository;
    private readonly ITagRepository _tagRepository;
    private readonly ISourceRepository _sourceRepository;

    public QuizService(IQuizRepository quizRepository, ITagRepository tagRepository, ISourceRepository sourceRepository)
    {
        _quizRepository = quizRepository;
        _tagRepository = tagRepository;
        _sourceRepository = sourceRepository;
    }

    public async Task<QuizDto?> GetQuizByIdAsync(int id, int? currentUserId)
    {
        // Fetch quiz even if disabled so that owner/admin can still see it
        var quiz = await _quizRepository.GetByIdAsync(id, includeDisabled: true);
        if (quiz == null)
            return null;

        // If disabled and not owner/admin, return null
        if (quiz.Disabled && quiz.UserId != currentUserId && !(currentUserId.HasValue && AdminHelper.IsAdmin(currentUserId.Value)))  // IsAdmin check would need to be passed or injected
            return null;

        // If private and not owner, return null
        if (quiz.Visibility == "private" && quiz.UserId != currentUserId)
            return null;

        return MapToDto(quiz);
    }

    public async Task<IEnumerable<QuizListDto>> GetUserQuizzesAsync(int userId)
    {
        var quizzes = await _quizRepository.GetUserQuizzesAsync(userId);
        return quizzes.Select(MapToListDto);
    }

    public async Task<IEnumerable<QuizListDto>> GetPublicQuizzesAsync(string? visibility = null)
    {
        var quizzes = await _quizRepository.GetPublicQuizzesAsync(visibility);
        return quizzes.Select(MapToListDto);
    }

    public async Task<QuizDto> CreateQuizAsync(int userId, CreateQuizDto dto)
    {
        var quiz = new Quiz
        {
            UserId = userId,
            Title = dto.Title,
            Description = dto.Description,
            Visibility = dto.Visibility ?? "private",
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        // Handle tags
        if (dto.TagNames != null && dto.TagNames.Any())
        {
            var tags = await _tagRepository.GetOrCreateTagsAsync(dto.TagNames);
            quiz.QuizTags = tags.Select(t => new QuizTag { Tag = t }).ToList();
        }

        var created = await _quizRepository.CreateAsync(quiz);
        return MapToDto(created);
    }

    public async Task<QuizDto?> UpdateQuizAsync(int id, int userId, UpdateQuizDto dto)
    {
        // Fetch quiz, including disabled if the user is admin
        var includeDisabled = AdminHelper.IsAdmin(userId);
        var quiz = await _quizRepository.GetByIdAsync(id, includeDisabled);
        if (quiz == null)
            return null;

        // If not admin, must be owner
        if (!AdminHelper.IsAdmin(userId) && quiz.UserId != userId)
            return null;

        quiz.Title = dto.Title ?? quiz.Title;
        quiz.Description = dto.Description ?? quiz.Description;
        quiz.Visibility = dto.Visibility ?? quiz.Visibility;

        // Admin can change disabled status
        if (AdminHelper.IsAdmin(userId) && dto.Disabled.HasValue)
        {
            quiz.Disabled = dto.Disabled.Value;
        }

        // Update tags if provided
        if (dto.TagNames != null)
        {
            // Remove existing tags
            await _tagRepository.RemoveQuizTags(quiz.Id);
            var tags = await _tagRepository.GetOrCreateTagsAsync(dto.TagNames);
            quiz.QuizTags = tags.Select(t => new QuizTag { QuizId = quiz.Id, TagId = t.Id }).ToList();
        }

        await _quizRepository.UpdateAsync(quiz);
        return MapToDto(quiz);
    }

    public async Task<bool> DeleteQuizAsync(int id, int userId, bool isAdmin)
    {
        // Fetch including disabled so admins can delete disabled quizzes too.
        var quiz = await _quizRepository.GetByIdAsync(id, includeDisabled: true);
        if (quiz == null)
            return false;

        if (!isAdmin && quiz.UserId != userId)
            return false;

        // Hard delete — removes the row and cascades to quiz_question,
        // quiz_attempt, quiz_attempt_answer via FK ON DELETE CASCADE.
        await _quizRepository.HardDeleteAsync(id);
        return true;
    }

    public async Task<QuizDto?> DuplicateQuizAsync(int id, int userId)
    {
        var original = await _quizRepository.GetByIdAsync(id);
        if (original == null)
            return null;

        // Only allow duplication if public or owned by user
        if (original.Visibility != "public" && original.UserId != userId)
            return null;

        var clone = new Quiz
        {
            UserId = userId,
            Title = $"{original.Title} (Copy)",
            Description = original.Description,
            Visibility = "private", // default private
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        // Clone tags
        if (original.QuizTags != null)
        {
            clone.QuizTags = original.QuizTags.Select(qt => new QuizTag { TagId = qt.TagId }).ToList();
        }

        // Clone questions (snapshots)
        if (original.QuizQuestions != null)
        {
            clone.QuizQuestions = original.QuizQuestions.Select(qq => new QuizQuestion
            {
                OriginalQuestionId = qq.OriginalQuestionId,
                QuestionSnapshotJson = qq.QuestionSnapshotJson,
                DisplayOrder = qq.DisplayOrder
            }).ToList();
        }

        // Clone sources?
        if (original.QuizSources != null)
        {
            clone.QuizSources = original.QuizSources.Select(qs => new QuizSource { SourceId = qs.SourceId }).ToList();
        }

        var created = await _quizRepository.CreateAsync(clone);
        return MapToDto(created);
    }

    public async Task<PaginatedResult<QuizListDto>> SearchQuizzesAsync(QuizSearchDto search, int page, int pageSize)
    {
        var (items, totalCount) = await _quizRepository.SearchAsync(
            search.SearchTerm,
            search.UserId,
            search.Visibility,
            search.IncludeDisabled,
            search.SortBy,
            search.SortOrder,
            page,
            pageSize);

        return new PaginatedResult<QuizListDto>
        {
            Items = items.Select(MapToListDto).ToList(),
            Total = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    private QuizDto MapToDto(Quiz quiz)
    {
        return new QuizDto
        {
            Id = quiz.Id,
            UserId = quiz.UserId,
            Title = quiz.Title,
            Description = quiz.Description,
            Visibility = quiz.Visibility,
            Disabled = quiz.Disabled,
            CreatedAt = quiz.CreatedAt,
            UpdatedAt = quiz.UpdatedAt,
            Tags = quiz.QuizTags?.Select(qt => qt.Tag?.Name ?? "").ToList() ?? new List<string>(),
            QuestionCount = quiz.QuizQuestions?.Count ?? 0,
            UserName = quiz.User?.Username
        };
    }

    private QuizListDto MapToListDto(Quiz quiz)
    {
        return new QuizListDto
        {
            Id = quiz.Id,
            Title = quiz.Title,
            Description = quiz.Description,
            UserId = quiz.UserId,
            Visibility = quiz.Visibility,
            Disabled = quiz.Disabled,
            CreatedAt = quiz.CreatedAt,
            UpdatedAt = quiz.UpdatedAt,
            Tags = quiz.QuizTags?.Select(qt => qt.Tag?.Name ?? "").ToList() ?? new List<string>(),
            QuestionCount = quiz.QuizQuestions?.Count ?? 0,
            UserName = quiz.User?.Username
        };
    }

    public async Task<IEnumerable<QuizQuestionDto>> GetQuizQuestionsAsync(int quizId, int? currentUserId)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            return new List<QuizQuestionDto>();

        // Check visibility
        if (quiz.Visibility == "private" && quiz.UserId != currentUserId)
            return new List<QuizQuestionDto>();

        var questions = await _quizRepository.GetQuestionsByQuizIdAsync(quizId);
        return questions.Select(MapToQuizQuestionDto);
    }

    public async Task<QuizQuestionDto?> GetQuizQuestionByIdAsync(int id, int? currentUserId)
    {
        var question = await _quizRepository.GetQuizQuestionByIdAsync(id);
        if (question == null || question.Quiz == null)
            return null;

        var quiz = question.Quiz;

        // If the quiz is disabled, treat as not found
        if (quiz.Disabled)
            return null;

        // Check visibility and ownership
        if (quiz.Visibility == "private" && quiz.UserId != currentUserId)
            return null;

        return MapToQuizQuestionDto(question);
    }

    public async Task<QuizQuestionDto> CreateQuizQuestionAsync(int quizId, int userId, CreateQuizQuestionDto dto)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            throw new ArgumentException("Quiz not found");

        if (quiz.UserId != userId)
            throw new UnauthorizedAccessException("You do not own this quiz");

        // Validate snapshot if provided
        if (!string.IsNullOrWhiteSpace(dto.QuestionSnapshotJson))
        {
            ValidateQuestionSnapshot(dto.QuestionSnapshotJson);
        }

        var question = new QuizQuestion
        {
            QuizId = quizId,
            OriginalQuestionId = dto.OriginalQuestionId,
            QuestionSnapshotJson = dto.QuestionSnapshotJson,
            DisplayOrder = dto.DisplayOrder
        };

        var created = await _quizRepository.CreateQuizQuestionAsync(question);
        return MapToQuizQuestionDto(created);
    }

    public async Task<QuizQuestionDto?> UpdateQuizQuestionAsync(int id, int userId, UpdateQuizQuestionDto dto)
    {
        var question = await _quizRepository.GetQuizQuestionByIdAsync(id);
        if (question == null)
            return null;

        if (question.Quiz.UserId != userId)
            throw new UnauthorizedAccessException("You do not own this quiz");

        // Validate snapshot if provided
        if (dto.QuestionSnapshotJson != null && !string.IsNullOrWhiteSpace(dto.QuestionSnapshotJson))
        {
            ValidateQuestionSnapshot(dto.QuestionSnapshotJson);
        }

        question.OriginalQuestionId = dto.OriginalQuestionId ?? question.OriginalQuestionId;
        question.QuestionSnapshotJson = dto.QuestionSnapshotJson ?? question.QuestionSnapshotJson;
        question.DisplayOrder = dto.DisplayOrder ?? question.DisplayOrder;
        question.UpdatedAt = DateTime.UtcNow;

        await _quizRepository.UpdateQuizQuestionAsync(question);
        return MapToQuizQuestionDto(question);
    }

    public async Task<bool> DeleteQuizQuestionAsync(int id, int userId, bool isAdmin)
    {
        var question = await _quizRepository.GetQuizQuestionByIdAsync(id);
        if (question == null)
            return false;

        if (!isAdmin && question.Quiz.UserId != userId)
            return false;

        await _quizRepository.DeleteQuizQuestionAsync(id);
        return true;
    }

    private QuizQuestionDto MapToQuizQuestionDto(QuizQuestion qq)
    {
        return new QuizQuestionDto
        {
            Id = qq.Id,
            QuizId = qq.QuizId,
            OriginalQuestionId = qq.OriginalQuestionId,
            QuestionSnapshotJson = qq.QuestionSnapshotJson,
            DisplayOrder = qq.DisplayOrder,
            CreatedAt = qq.CreatedAt,
            UpdatedAt = qq.UpdatedAt
        };
    }

    // ------------------------------------------------------------------------
    // NEW: Validate question snapshot using the shared helper
    // ------------------------------------------------------------------------
    private void ValidateQuestionSnapshot(string snapshotJson)
    {
        try
        {
            using JsonDocument doc = JsonDocument.Parse(snapshotJson);
            JsonElement root = doc.RootElement;

            // Extract required fields
            if (!root.TryGetProperty("type", out JsonElement typeElem) || typeElem.ValueKind != JsonValueKind.String)
                throw new ArgumentException("Question snapshot must contain a 'type' field (string).");
            string type = typeElem.GetString()!;

            if (!root.TryGetProperty("content", out JsonElement contentElem) || contentElem.ValueKind != JsonValueKind.String)
                throw new ArgumentException("Question snapshot must contain a 'content' field (string).");
            string content = contentElem.GetString()!;

            // Metadata may be an object or null; convert to JSON string for validation
            string metadataJson = null;
            if (root.TryGetProperty("metadata", out JsonElement metadataElem))
            {
                metadataJson = metadataElem.ValueKind == JsonValueKind.Null ? "null" : metadataElem.GetRawText();
            }

            // Validate using the shared helper
            QuestionValidationHelper.ValidateQuestionMetadata(type, content, metadataJson);
        }
        catch (JsonException ex)
        {
            throw new ArgumentException("Invalid JSON format in question snapshot.", ex);
        }
    }

    public async Task<IEnumerable<NoteDto>> GetQuizNotesAsync(int quizId, int? currentUserId)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            return Enumerable.Empty<NoteDto>();

        // Check visibility
        if (quiz.Visibility == "private" && quiz.UserId != currentUserId)
            return Enumerable.Empty<NoteDto>();

        var notes = await _quizRepository.GetNotesByQuizIdAsync(quizId);
        return notes.Select(n => new NoteDto
        {
            Id = n.Id,
            Title = n.Title,
            Content = n.Content,
            CreatedAt = n.CreatedAt,
            UpdatedAt = n.UpdatedAt
        });
    }

    public async Task<NoteDto> AddNoteToQuizAsync(int quizId, int userId, AddNoteToQuizDto dto)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            throw new ArgumentException("Quiz not found");

        if (quiz.UserId != userId)
            throw new UnauthorizedAccessException("You do not own this quiz");

        Note.Note? note;
        if (dto.NoteId.HasValue)
        {
            // Attach existing note – verify it belongs to the user
            note = await _quizRepository.GetNoteByIdAndUserAsync(dto.NoteId.Value, userId);
            if (note == null)
                throw new ArgumentException("Note not found or does not belong to you");

            await _quizRepository.AddNoteToQuizAsync(quizId, note.Id);
        }
        else if (!string.IsNullOrWhiteSpace(dto.Title) && !string.IsNullOrWhiteSpace(dto.Content))
        {
            // Create new note
            note = await _quizRepository.CreateNoteAndAddToQuizAsync(quizId, userId, dto.Title, dto.Content);
        }
        else
        {
            throw new ArgumentException("Either provide a NoteId or both Title and Content");
        }

        return new NoteDto
        {
            Id = note.Id,
            Title = note.Title,
            Content = note.Content,
            CreatedAt = note.CreatedAt,
            UpdatedAt = note.UpdatedAt
        };
    }

    public async Task<bool> RemoveNoteFromQuizAsync(int quizId, int noteId, int userId, bool isAdmin)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            return false;

        if (!isAdmin && quiz.UserId != userId)
            return false;

        await _quizRepository.RemoveNoteFromQuizAsync(quizId, noteId);
        return true;
    }

    public async Task<IEnumerable<SourceDto>> GetQuizSourcesAsync(int quizId, int? currentUserId)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            return Enumerable.Empty<SourceDto>();

        // Respect privacy: only owner can see sources of private quizzes
        if (quiz.Visibility == "private" && quiz.UserId != currentUserId)
            return Enumerable.Empty<SourceDto>();

        var sources = await _quizRepository.GetSourcesByQuizIdAsync(quizId);
        return sources.Select(MapSourceToDto);
    }

    public async Task<SourceDto> AddSourceToQuizAsync(int quizId, int userId, int sourceId)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            throw new ArgumentException("Quiz not found");

        if (quiz.UserId != userId)
            throw new UnauthorizedAccessException("You do not own this quiz");

        var source = await _sourceRepository.GetByIdAsync(sourceId);
        if (source == null || source.UserId != userId)
            throw new ArgumentException("Source not found or does not belong to you");

        await _quizRepository.AddSourceToQuizAsync(quizId, sourceId);
        return MapSourceToDto(source);
    }

    public async Task<bool> RemoveSourceFromQuizAsync(int quizId, int sourceId, int userId, bool isAdmin)
    {
        var quiz = await _quizRepository.GetByIdAsync(quizId);
        if (quiz == null)
            return false;

        if (!isAdmin && quiz.UserId != userId)
            return false;

        await _quizRepository.RemoveSourceFromQuizAsync(quizId, sourceId);
        return true;
    }

    // Helper mapping method (same as in SourceService)
    private SourceDto MapSourceToDto(Source.Source s)
    {
        return new SourceDto
        {
            Id = s.Id,
            UserId = s.UserId,
            Type = s.Type,
            Title = s.Title,
            Url = s.Url,
            RawText = s.RawText,
            FilePath = s.FilePath,
            CreatedAt = s.CreatedAt,
            UpdatedAt = s.UpdatedAt,
            Metadata = s.MetadataJson
        };
    }
}

File: slp/backend-nestjs/src/modules/quiz/quiz.service.ts
import {
  Injectable,
  NotFoundException,
  UnauthorizedException,
  BadRequestException,
} from "@nestjs/common";
import { QuizRepository } from "./quiz.repository";
import { TagRepository } from "../tag/tag.repository";
import { SourceRepository } from "../source/source.repository";
import { Quiz } from "./quiz.entity";
import { QuizDto, QuizListDto } from "./dto/quiz.dto";
import { CreateQuizDto } from "./dto/create-quiz.dto";
import { UpdateQuizDto } from "./dto/update-quiz.dto";
import {
  CreateQuizQuestionDto,
  UpdateQuizQuestionDto,
  QuizQuestionDto,
} from "./dto/quiz-question.dto";
import { AddNoteToQuizDto } from "./dto/add-note-to-quiz.dto";
import { AddSourceToQuizDto } from "./dto/add-source-to-quiz.dto";
import { NoteDto } from "../note/dto/note.dto";
import { SourceDto } from "src/modules/source/source.dto";
import { QuizSearchDto } from "./dto/quiz-search.dto";
import { PaginatedResult } from "../../helpers/pagination.helper";
import { isAdmin } from "../../helpers/admin.helper";
import { validate as validateJson } from "class-validator";
import { QuestionValidationHelper } from "../question/question-validation.helper"; // if exists
import { QuizTag } from "./quiz-tag.entity";
import { QuizQuestion } from "./quiz-question.entity";
import { QuizSource } from "./quiz-source.entity";

@Injectable()
export class QuizService {
  constructor(
    private quizRepo: QuizRepository,
    private tagRepo: TagRepository,
    private sourceRepo: SourceRepository,
  ) {}

  private mapToDto(quiz: Quiz): QuizDto {
    return {
      id: quiz.id,
      userId: quiz.userId,
      title: quiz.title,
      description: quiz.description,
      visibility: quiz.visibility,
      disabled: quiz.disabled,
      createdAt: quiz.createdAt,
      updatedAt: quiz.updatedAt,
      tags: quiz.quizTags?.map((qt) => qt.tag?.name) ?? [],
      questionCount: quiz.quizQuestions?.length ?? 0,
      userName: quiz.user?.username,
    };
  }

  private mapToListDto(quiz: Quiz): QuizListDto {
    return {
      id: quiz.id,
      title: quiz.title,
      description: quiz.description,
      userId: quiz.userId,
      visibility: quiz.visibility,
      disabled: quiz.disabled,
      createdAt: quiz.createdAt,
      updatedAt: quiz.updatedAt,
      tags: quiz.quizTags?.map((qt) => qt.tag?.name) ?? [],
      questionCount: quiz.quizQuestions?.length ?? 0,
      userName: quiz.user?.username,
    };
  }

  private mapToQuestionDto(question: any): QuizQuestionDto {
    return {
      id: question.id,
      quizId: question.quizId,
      originalQuestionId: question.originalQuestionId,
      questionSnapshotJson: question.questionSnapshotJson,
      displayOrder: question.displayOrder,
      createdAt: question.createdAt,
      updatedAt: question.updatedAt,
    };
  }

  private validateQuestionSnapshot(snapshotJson: string): void {
    try {
      const doc = JSON.parse(snapshotJson);
      if (!doc.type || typeof doc.type !== "string") {
        throw new BadRequestException(
          'Question snapshot must contain a "type" field (string).',
        );
      }
      if (!doc.content || typeof doc.content !== "string") {
        throw new BadRequestException(
          'Question snapshot must contain a "content" field (string).',
        );
      }
      // Optionally validate using QuestionValidationHelper if available
      // QuestionValidationHelper.validateQuestionMetadata(doc.type, doc.content, doc.metadata);
    } catch (err) {
      throw new BadRequestException(
        "Invalid JSON format in question snapshot.",
      );
    }
  }

  async getQuizById(
    id: number,
    currentUserId?: number,
  ): Promise<QuizDto | null> {
    const quiz = await this.quizRepo.getById(id, true); // include disabled
    if (!quiz) return null;

    // If disabled and not owner/admin, return null
    if (
      quiz.disabled &&
      quiz.userId !== currentUserId &&
      !(currentUserId && isAdmin(currentUserId))
    ) {
      return null;
    }

    // If private and not owner
    if (quiz.visibility === "private" && quiz.userId !== currentUserId) {
      return null;
    }

    return this.mapToDto(quiz);
  }

  async getUserQuizzes(userId: number): Promise<QuizListDto[]> {
    const quizzes = await this.quizRepo.getUserQuizzes(userId);
    return quizzes.map(this.mapToListDto);
  }

  async getPublicQuizzes(visibility?: string): Promise<QuizListDto[]> {
    const quizzes = await this.quizRepo.getPublicQuizzes(visibility);
    return quizzes.map(this.mapToListDto);
  }

async createQuiz(userId: number, dto: CreateQuizDto): Promise<QuizDto> {
    const quiz = new Quiz();
    quiz.userId = userId;
    quiz.title = dto.title;
    quiz.description = dto.description ?? null;
    quiz.visibility = dto.visibility ?? "private";
    quiz.createdAt = new Date();
    quiz.updatedAt = new Date();

    if (dto.tagNames?.length) {
      const tags = await this.tagRepo.getOrCreateTags(dto.tagNames);
      quiz.quizTags = tags.map((tag) => {
        const qt = new QuizTag();
        qt.tag = tag;
        return qt;
      });
    }

    const created = await this.quizRepo.create(quiz);
    return this.mapToDto(created);
  }

  async updateQuiz(
    id: number,
    userId: number,
    dto: UpdateQuizDto,
  ): Promise<QuizDto | null> {
    const includeDisabled = isAdmin(userId);
    const quiz = await this.quizRepo.getById(id, includeDisabled);
    if (!quiz) return null;

    if (!isAdmin(userId) && quiz.userId !== userId) return null;

    if (dto.title !== undefined) quiz.title = dto.title;
    if (dto.description !== undefined) quiz.description = dto.description;
    if (dto.visibility !== undefined) quiz.visibility = dto.visibility;
    if (isAdmin(userId) && dto.disabled !== undefined)
      quiz.disabled = dto.disabled;

    if (dto.tagNames !== undefined) {
      // Remove existing tags
      await this.quizRepo.removeQuizTags(quiz.id);
      // Create new tags and associate
      const tags = await this.tagRepo.getOrCreateTags(dto.tagNames);
      quiz.quizTags = tags.map((tag) => {
        const qt = new QuizTag();
        qt.quizId = quiz.id;
        qt.tagId = tag.id;
        qt.tag = tag;
        return qt;
      });
    }

    await this.quizRepo.update(quiz);
    return this.mapToDto(quiz);
  }

  async deleteQuiz(
    id: number,
    userId: number,
    isAdmin: boolean,
  ): Promise<boolean> {
    const quiz = await this.quizRepo.getById(id, true);
    if (!quiz) return false;
    if (!isAdmin && quiz.userId !== userId) return false;
    await this.quizRepo.hardDelete(id);
    return true;
  }

  async duplicateQuiz(id: number, userId: number): Promise<QuizDto | null> {
    const original = await this.quizRepo.getById(id);
    if (!original) return null;
    if (original.visibility !== "public" && original.userId !== userId)
      return null;

    const clone = new Quiz();
    clone.userId = userId;
    clone.title = `${original.title} (Copy)`;
    clone.description = original.description;
    clone.visibility = "private";
    clone.createdAt = new Date();
    clone.updatedAt = new Date();

    if (original.quizTags?.length) {
      clone.quizTags = original.quizTags.map((qt) => {
        const newQt = new QuizTag();
        newQt.tagId = qt.tagId;
        return newQt;
      });
    }

    if (original.quizQuestions?.length) {
      clone.quizQuestions = original.quizQuestions.map((qq) => {
        const newQq = new QuizQuestion();
        newQq.originalQuestionId = qq.originalQuestionId;
        newQq.questionSnapshotJson = qq.questionSnapshotJson;
        newQq.displayOrder = qq.displayOrder;
        return newQq;
      });
    }

    if (original.quizSources?.length) {
      clone.quizSources = original.quizSources.map((qs) => {
        const newQs = new QuizSource();
        newQs.sourceId = qs.sourceId;
        return newQs;
      });
    }

    const created = await this.quizRepo.create(clone);
    return this.mapToDto(created);
  }

  async createQuizQuestion(
    quizId: number,
    userId: number,
    dto: CreateQuizQuestionDto,
  ): Promise<QuizQuestionDto> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) throw new NotFoundException("Quiz not found");
    if (quiz.userId !== userId)
      throw new UnauthorizedException("You do not own this quiz");
    if (dto.questionSnapshotJson) {
      this.validateQuestionSnapshot(dto.questionSnapshotJson);
    }

    const question = new QuizQuestion();
    question.quizId = quizId;
    question.originalQuestionId = dto.originalQuestionId ?? null;
    question.questionSnapshotJson = dto.questionSnapshotJson ?? null;
    question.displayOrder = dto.displayOrder;

    const created = await this.quizRepo.createQuizQuestion(question);
    return this.mapToQuestionDto(created);
  }

  async updateQuizQuestion(
    id: number,
    userId: number,
    dto: UpdateQuizQuestionDto,
  ): Promise<QuizQuestionDto | null> {
    const question = await this.quizRepo.getQuizQuestionById(id);
    if (!question) return null;
    if (question.quiz.userId !== userId)
      throw new UnauthorizedException("You do not own this quiz");
    if (dto.questionSnapshotJson) {
      this.validateQuestionSnapshot(dto.questionSnapshotJson);
    }

    if (dto.originalQuestionId !== undefined)
      question.originalQuestionId = dto.originalQuestionId;
    if (dto.questionSnapshotJson !== undefined)
      question.questionSnapshotJson = dto.questionSnapshotJson;
    if (dto.displayOrder !== undefined)
      question.displayOrder = dto.displayOrder;
    question.updatedAt = new Date();

    await this.quizRepo.updateQuizQuestion(question);
    return this.mapToQuestionDto(question);
  }

  async searchQuizzes(
    search: QuizSearchDto,
    page: number,
    pageSize: number,
  ): Promise<PaginatedResult<QuizListDto>> {
    const { items, totalCount } = await this.quizRepo.search(
      search.searchTerm,
      search.userId,
      search.visibility,
      search.includeDisabled,
      search.sortBy,
      search.sortOrder,
      page,
      pageSize,
    );
    return {
      items: items.map(this.mapToListDto),
      total: totalCount,
      page,
      pageSize,
      totalPages: Math.ceil(totalCount / pageSize),
    };
  }

  async getQuizQuestions(
    quizId: number,
    currentUserId?: number,
  ): Promise<QuizQuestionDto[]> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) return [];
    if (quiz.visibility === "private" && quiz.userId !== currentUserId)
      return [];
    const questions = await this.quizRepo.getQuestionsByQuizId(quizId);
    return questions.map(this.mapToQuestionDto);
  }

  async getQuizQuestionById(
    id: number,
    currentUserId?: number,
  ): Promise<QuizQuestionDto | null> {
    const question = await this.quizRepo.getQuizQuestionById(id);
    if (!question || !question.quiz) return null;
    if (question.quiz.disabled) return null;
    if (
      question.quiz.visibility === "private" &&
      question.quiz.userId !== currentUserId
    )
      return null;
    return this.mapToQuestionDto(question);
  }

  async deleteQuizQuestion(
    id: number,
    userId: number,
    isAdmin: boolean,
  ): Promise<boolean> {
    const question = await this.quizRepo.getQuizQuestionById(id);
    if (!question) return false;
    if (!isAdmin && question.quiz.userId !== userId) return false;
    await this.quizRepo.deleteQuizQuestion(id);
    return true;
  }

  async getQuizNotes(
    quizId: number,
    currentUserId?: number,
  ): Promise<NoteDto[]> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) return [];
    if (quiz.visibility === "private" && quiz.userId !== currentUserId)
      return [];
    const notes = await this.quizRepo.getNotesByQuizId(quizId);
    return notes.map((n) => ({
      id: n.id,
      title: n.title,
      content: n.content,
      createdAt: n.createdAt,
      updatedAt: n.updatedAt,
    }));
  }

  async addNoteToQuiz(
    quizId: number,
    userId: number,
    dto: AddNoteToQuizDto,
  ): Promise<NoteDto> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) throw new NotFoundException("Quiz not found");
    if (quiz.userId !== userId)
      throw new UnauthorizedException("You do not own this quiz");

    let note;
    if (dto.noteId) {
      note = await this.quizRepo.getNoteByIdAndUser(dto.noteId, userId);
      if (!note)
        throw new BadRequestException(
          "Note not found or does not belong to you",
        );
      await this.quizRepo.addNoteToQuiz(quizId, note.id);
    } else if (dto.title && dto.content) {
      note = await this.quizRepo.createNoteAndAddToQuiz(
        quizId,
        userId,
        dto.title,
        dto.content,
      );
    } else {
      throw new BadRequestException(
        "Either provide a NoteId or both Title and Content",
      );
    }

    return {
      id: note.id,
      title: note.title,
      content: note.content,
      createdAt: note.createdAt,
      updatedAt: note.updatedAt,
    };
  }

  async removeNoteFromQuiz(
    quizId: number,
    noteId: number,
    userId: number,
    isAdmin: boolean,
  ): Promise<boolean> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) return false;
    if (!isAdmin && quiz.userId !== userId) return false;
    await this.quizRepo.removeNoteFromQuiz(quizId, noteId);
    return true;
  }

  async getQuizSources(
    quizId: number,
    currentUserId?: number,
  ): Promise<SourceDto[]> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) return [];
    if (quiz.visibility === "private" && quiz.userId !== currentUserId)
      return [];
    const sources = await this.quizRepo.getSourcesByQuizId(quizId);
    return sources.map((s) => ({
      id: s.id,
      userId: s.userId,
      type: s.type,
      title: s.title,
      url: s.url,
      rawText: s.rawText,
      filePath: s.filePath,
      createdAt: s.createdAt,
      updatedAt: s.updatedAt,
      metadata: s.metadataJson,
    }));
  }

  async addSourceToQuiz(
    quizId: number,
    userId: number,
    sourceId: number,
  ): Promise<SourceDto> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) throw new NotFoundException("Quiz not found");
    if (quiz.userId !== userId)
      throw new UnauthorizedException("You do not own this quiz");

    const source = await this.sourceRepo.getById(sourceId);
    if (!source || source.userId !== userId)
      throw new BadRequestException(
        "Source not found or does not belong to you",
      );

    await this.quizRepo.addSourceToQuiz(quizId, sourceId);
    return {
      id: source.id,
      userId: source.userId,
      type: source.type,
      title: source.title,
      url: source.url,
      rawText: source.rawText,
      filePath: source.filePath,
      createdAt: source.createdAt,
      updatedAt: source.updatedAt,
      metadata: source.metadataJson,
    };
  }

  async removeSourceFromQuiz(
    quizId: number,
    sourceId: number,
    userId: number,
    isAdmin: boolean,
  ): Promise<boolean> {
    const quiz = await this.quizRepo.getById(quizId);
    if (!quiz) return false;
    if (!isAdmin && quiz.userId !== userId) return false;
    await this.quizRepo.removeSourceFromQuiz(quizId, sourceId);
    return true;
  }
}
